File: app/src/main/java/com/microblocklabs/mpc/interceptor/GrpcClientRequestInterceptor.java
package com.microblocklabs.mpc.interceptor;

import io.grpc.CallOptions;
import io.grpc.Channel;
import io.grpc.ClientCall;
import io.grpc.ClientInterceptor;
import io.grpc.ForwardingClientCall;
import io.grpc.Metadata;
import io.grpc.MethodDescriptor;

public class GrpcClientRequestInterceptor implements ClientInterceptor {
    @Override
    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
            final MethodDescriptor<ReqT, RespT> methodDescriptor,
            final CallOptions callOptions,
            final Channel channel) {
        return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
                channel.newCall(methodDescriptor, callOptions)) {
            @Override
            public void start(Listener<RespT> responseListener, Metadata headers) {
//                var userToken = UserContext.getUserContext().getUserToken();
////                log.info("Setting userToken in header");
//                headers.put(Metadata.Key.of("JWT", Metadata.ASCII_STRING_MARSHALLER), userToken);
                super.start(responseListener, headers);
            }

            @Override
            public void sendMessage(ReqT message) {
                super.sendMessage(message);
            }
        };
    }
}

File: app/src/main/java/com/microblocklabs/mpc/interceptor/GrpcClientRequestInterceptor.kt
package com.microblocklabs.mpc.interceptor

import android.util.Log
import aot.armsproject.utils.AppSharedPreferenceManager
import com.microblocklabs.mpc.utility.Constant
import io.grpc.CallOptions
import io.grpc.Channel
import io.grpc.ClientCall
import io.grpc.ClientInterceptor
import io.grpc.ForwardingClientCall.SimpleForwardingClientCall
import io.grpc.Metadata
import io.grpc.MethodDescriptor


class GrpcClientRequestInterceptor(mpcSharedPref: AppSharedPreferenceManager) : ClientInterceptor {
    val token =  mpcSharedPref.getString(Constant.SessionToken)
    override fun <ReqT, RespT> interceptCall(
        methodDescriptor: MethodDescriptor<ReqT, RespT>?,
        callOptions: CallOptions?,
        channel: Channel
    ): ClientCall<ReqT, RespT> {
        return object : SimpleForwardingClientCall<ReqT, RespT>(channel.newCall(methodDescriptor, callOptions)) {
            override fun start(responseListener: Listener<RespT>?, headers: Metadata) {
                val userToken = "Bearer $token"
                headers.put(Metadata.Key.of(ACCESS_TOKEN, Metadata.ASCII_STRING_MARSHALLER), userToken)
                super.start(responseListener, headers)
            }
        }
    }

    companion object {
        private const val ACCESS_TOKEN = "authorization"
    }
}